File: go_code/server/storage/redis_store.go
package storage

import (
	"context"
	"encoding/json"
	"time"

	"github.com/redis/go-redis/v9"
)

// RedisStore provides Redis-based key-value storage
type RedisStore struct {
	redisClient *redis.Client
}

// NewRedisStore creates a new Redis store
func NewRedisStore(addr string) *RedisStore {
	client := redis.NewClient(&redis.Options{
		Addr: addr,
	})
	return &RedisStore{
		redisClient: client,
	}
}

// Store stores a key-value pair with optional TTL
func (s *RedisStore) Store(key string, value interface{}, ttl *time.Duration) bool {
	ctx := context.Background()

	// Serialize value to JSON
	jsonValue, err := json.Marshal(value)
	if err != nil {
		return false
	}

	if ttl != nil {
		err = s.redisClient.Set(ctx, key, jsonValue, *ttl).Err()
	} else {
		err = s.redisClient.Set(ctx, key, jsonValue, 0).Err()
	}

	return err == nil
}

// Retrieve retrieves a value by key
func (s *RedisStore) Retrieve(key string) interface{} {
	ctx := context.Background()

	result, err := s.redisClient.Get(ctx, key).Result()
	if err != nil {
		return nil
	}

	// Try to unmarshal as JSON first
	var value interface{}
	if err := json.Unmarshal([]byte(result), &value); err != nil {
		// If JSON unmarshaling fails, return as string
		return result
	}

	return value
}

// Delete deletes a key-value pair
func (s *RedisStore) Delete(key string) bool {
	ctx := context.Background()

	result, err := s.redisClient.Del(ctx, key).Result()
	return err == nil && result > 0
}

// Exists checks if a key exists
func (s *RedisStore) Exists(key string) bool {
	ctx := context.Background()

	result, err := s.redisClient.Exists(ctx, key).Result()
	return err == nil && result > 0
}

File: server/storage/redis_store.py
import redis
import json
from typing import Any, List, Optional
from storage.key_value_store import KeyValueStore


"""
Redis store for the key-value store
"""
class RedisStore(KeyValueStore):
    def __init__(self, redis_client: redis.Redis):
        self.redis_client = redis_client
    
    def _serialize_value(self, value: Any) -> str:
        return json.dumps(value)
    
    def _deserialize_value(self, value: str) -> Any:
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return value
    
    async def store(self, key: str, value: Any, ttl: Optional[int] = None) -> bool:
        try:
            serialized_value = self._serialize_value(value)
            if ttl:
                self.redis_client.setex(key, ttl, serialized_value)
            else:
                self.redis_client.set(key, serialized_value)
            return True
        except redis.RedisError:
            return False
    
    async def retrieve(self, key: str) -> Optional[Any]:
        try:
            value = self.redis_client.get(key)
            if value is None:
                return None
            return self._deserialize_value(value)
        except redis.RedisError:
            return None
    
    async def delete(self, key: str) -> bool:
        try:
            result = self.redis_client.delete(key)
            return result > 0
        except redis.RedisError:
            return False
    
    async def exists(self, key: str) -> bool:
        try:
            return bool(self.redis_client.exists(key))
        except redis.RedisError:
            return False